core: pager memory optim in rpmb fs driver

etienne-lms opened the pull request:
RPMB FS driver may allocates a temporary buffer in the heap on caller data transfer request. These may be big buffers of dozens of kbytes. When pager is enabled and heap not so big such temporary buffers could be allocated in fully pageable read/write memory, preserving the heap.

This P-R addresses that.

jenswi-linaro reviewed (commented):
This seems to try to solve the same problem as with
https://github.com/OP-TEE/optee_os/blob/3f66770601fa81b2dd16a51d66edf3030153da07/core/tee/tee_ree_fs.c#L53-L61

Could that approach be used here too?

etienne-lms commented on core/tee/tee_rpmb_fs.c:
```diff
@@ -2369,6 +2372,92 @@ static TEE_Result rpmb_fs_read(struct tee_file_handle *tfh, size_t pos,
+ * Register in pager a pageable read/write memory range
+ * used as a heap from Core when prefer pageable read/write
+ * buffers over the standard heap due to its constrainted size.
+ * 2 extra pages are added and not mapped as boundaries protection
```
Sorry I removed this part. If useful, it should be done from `tee_pager_alloc()`. I have some change proposal for that after this P-R. I'll remove this inline comment.

etienne-lms commented:
> This seems to try to solve the same problem as with
Could that approach be used here too?

Sure :)  I'll look at it and consider the pager config.

etienne-lms commented:
updated: moved rpmb fs to default mempool and make default mempool pageable memory.
I wonder in the mempool for userland sized buffer would deserve a `CFG_xxx_SIZE` config switch?

jenswi-linaro commented:
Please fixup and squash the commits. I think we need to decide how large writes we're going to support, setting it to the size of the heap seems rather arbitrary to me.

> I wonder in the mempool for userland sized buffer would deserve a `CFG_xxx_SIZE` config switch?

Hmm, maybe it should be a per TA config option instead?

etienne-lms commented:
squashed

jenswi-linaro commented:
> squashed

Thanks.

I think we need to decide how large writes we're going to support, setting it to the size of the heap seems rather arbitrary to me. What is the advantage with using `CFG_CORE_HEAP_SIZE` over `MPI_MEMPOOL_SIZE` where the latter is hard coded to 42kB? An obvious drawback is that's it more complicated due to the two extra commits. Another drawback (which could be fixed) is that it's only maybe shared with the bignum memory pool depending on pager config.

etienne-lms commented:
Maybe always move `mempool_default` out of MPI scratch mem.

`mempool_default` is used only by ree-fs and rpmb-fs drivers for temporary buffers. We could replace `mempool_default` with a `user_buffer_mempool`  using the same scheme as for mpi scratchmem:
w/ pager => `tee_pager_alloc()` (but pager unlocked r/w vmem).
w/o pager => static aligned byte arrays from zi/bss section.

As for the size, maybe `CFG_USER_BUFFER_MEMPOOL_SIZE`

jenswi-linaro commented:
> Maybe always move `mempool_default` out of MPI scratch mem.

Make sense.

> 
> `mempool_default` is used only by ree-fs and rpmb-fs drivers for temporary buffers. We could replace `mempool_default` with a `user_buffer_mempool` using the same scheme as for mpi scratchmem:
> w/ pager => `tee_pager_alloc()` (but pager unlocked r/w vmem).
> w/o pager => static aligned byte arrays from zi/bss section.
> 
> As for the size, maybe `CFG_USER_BUFFER_MEMPOOL_SIZE`

Looking at this critical part of `rpmb_fs_write_primitive()` again, it seems the implementation is broken since it assumes that we have enough memory to hold the entire file in memory. The part that needs to be atomic is the part updating the fat entry. The file  could be copied into a the new location piece by piece. It wouldn't hurt from a performance point of view to be able to copy a few kB:s at a time, but we don't need any extreme sizes. Am I missing something?

etienne-lms commented:
I'm closing the P-R.
Superseded by https://github.com/OP-TEE/optee_os/pull/4299 that only changes rpmb fs implementation to use a temporary buffer from the default mempool.
